src/webots/core/WbNetwork.cpp

```diff
@@ -93,6 +102,103 @@ void WbNetwork::setProxy() {
     mNetworkAccessManager->setProxy(proxy);
 }
 
+void WbNetwork::save(const QString &url, const QByteArray &content) {
+  if (!isCached(url)) {
+    // save to file
+    const QString path = mCacheDirectory + urlToHash(url);
+    QFile file(path);
+    if (file.open(QIODevice::WriteOnly)) {
+      file.write(content);
+      mCacheSizeInBytes += file.size();
+      file.close();
+      // save reference in internal representation
+      cacheMap.insert(url, path);
+    }
+  }
+}
+
+const QString WbNetwork::get(const QString &url) const {
+  // cppcheck-suppress assertWithSideEffect
+  assert(isCached(url));  // the 'get' function should not be called unless we know that the file is cached
+
+  if (cacheMap.contains(url))
+    return cacheMap[url];
+
+  const QString location = mCacheDirectory + urlToHash(url);
+  cacheMap.insert(url, location);
+
+  return location;
```

Comment:
Is it possible to reach this code ? 
The `assert(isCached(url))` return true only if `cacheMap` already contains the url or if the file exists on the disk and in the latest case, it inserts it in the `cacheMap`.
So if we pass the assert, the test at line 124 will always pass?

Reply:
The assert is a sanity check when developing. If it's triggered it means we didn't implement the retrieval logic correctly, in other words we shouldn't have reached to point of calling the get function unless the file is in the local cache.

`cacheMap` is not a complete representation of the state of the cache, it's a representation that we build progressively as assets are requested/returned, and the information is lost when closing Webots. 
By keeping this internal representation allows to skip having to compute the hash from the url and checking if the file exists on the disk every time, which otherwise we would do even when the same url was requested before (or rather, in the current session).